### PR TITLE
tiles: add gossip tile

### PR DESCRIPTION
**Pull request by lheeger-jump**

### CODE REVIEW

**Review by arjain4 — CHANGES_REQUESTED** [review]

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof(fd_gossip_tile_ctx_t) );
+}
+typedef struct __attribute__((packed)) {
+  fd_eth_hdr_t eth[1];
+  fd_ip4_hdr_t ip4[1];
+  fd_udp_hdr_t udp[1];
+} eth_ip_udp_t;
+
+static inline void
```
[the remark this answers is not in the document]
Comment (lheeger-jump) [reply src/app/fdctl/run/tiles/fd_gossip.c:217]: Yes we can do this in follow up

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+static void
+during_frag( void * _ctx,
+             ulong  in_idx      FD_PARAM_UNUSED,
+             ulong  seq         FD_PARAM_UNUSED,
+             ulong  sig         FD_PARAM_UNUSED,
+             ulong  chunk,
+             ulong  sz,
+             int *  opt_filter ) {
+  fd_gossip_tile_ctx_t * ctx = (fd_gossip_tile_ctx_t *) _ctx;
+
```
[the remark this answers is not in the document]
Comment (lheeger-jump) [reply src/app/fdctl/run/tiles/fd_gossip.c:352]: done

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+#define SIGN_IN_IDX     1
+
+#define SHRED_OUT_IDX   0
+#define REPAIR_OUT_IDX  1
+#define PACK_OUT_IDX    2
+#define SIGN_OUT_IDX    3
+
+#define CONTACT_INFO_PUBLISH_TIME_NS ((long)5e9)
+
+struct __attribute__((packed)) fd_shred_dest_wire {
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_gossip.c:35]: These have been shifted out to util in the repair PR, just as a note

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+  ulong gossip_buffer_sz;
+  uchar gossip_buffer[ FD_NET_MTU ];
+
+  uchar src_mac_addr[6];
+
+  fd_keyguard_client_t  keyguard_client[1];
+};
+typedef struct fd_gossip_tile_ctx fd_gossip_tile_ctx_t;
+
+
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_gossip.c:146]: nit: random whitespace

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+    return NULL;
+  }
+  res->port = htons( (ushort)port );
+
+  return res;
+}
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return 4096UL;
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_gossip.c:189]: is 4096 required? Might work with 128

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+
+  uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->net_in, chunk );
+
+  ctx->gossip_buffer_sz = sz;
+  fd_memcpy( ctx->gossip_buffer, dcache_entry, sz );
+}
+
+static void
+after_frag( void *             _ctx,
+            ulong              in_idx,
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_gossip.c:372]: FD_PARAM_UNUSED is the way to go it turns out :)

File: src/app/fdctl/run/tiles/fd_gossip.c [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+  ulong out_cnt = 0;
+  out_fds[ out_cnt++ ] = 2; /* stderr */
+  if( FD_LIKELY( -1!=fd_log_private_logfile_fd() ) )
+    out_fds[ out_cnt++ ] = fd_log_private_logfile_fd(); /* logfile */
+  return out_cnt;
+}
+
+fd_topo_run_tile_t fd_tile_gossip = {
+  .name                     = "gossip",
+  .mux_flags                = FD_MUX_FLAG_COPY,
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_gossip.c:710]: Does this need a manual publish flag? My understanding is that we don't need to publish every time

**Review by arjain4 — CHANGES_REQUESTED** [review]
Final changes i think

**Review by asiegel-jt — APPROVED** [review]

File: src/disco/topo/fd_topo.h [hunk 8]
```diff
@@ -210,6 +211,27 @@ typedef struct {
     struct {
       ulong accounts_cnt;
     } benchg;
+
+    struct {
+      char    gossip_peer_addr[ 22 ]; // len('255.255.255.255:65535') == 22
```
Comment (mmcgee-jump) [inline src/disco/topo/fd_topo.h:216]: uint gossip_peer_ip_addr;

**Pull request closed without merging**